### PR TITLE
AWS EC2Autoscaler -- Accept Region

### PR DESCRIPTION
**Pull request by nicktylah**
- Update overlord autoscaling config to contain "region" key, which will
  inform EC2Autoscaler's amazonEC2Client.
- Set the appropriate region (if not null) in `AWSModule`'s
  `getEC2Client` method
- aws-java-sdk defaults to `us-east-1`, preventing the autoscaler from
  spinning up instances outside of that region (e.g. `us-west-2a`)

### CODE REVIEW

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java [hunk 1]
```diff
@@ -70,6 +72,13 @@ public EC2AutoScaler(
     this.envConfig = envConfig;
     this.amazonEC2Client = amazonEC2Client;
     this.config = config;
+
+    // Set the AWS region to work in according to envConfig, if set.
+    try {
+      amazonEC2Client.setRegion(RegionUtils.getRegion(envConfig.getRegion()));
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java]: Suggest taking region from config and simply skipping this step if it is null. Doing so retains prior behavior.


```java
final String region = envConfig.getRegion();
if(region != null) {
  amazonEC2Client.setRegion(RegionUtils.getRegion(region));
}
```

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java [hunk 2]
```diff
@@ -70,6 +72,13 @@ public EC2AutoScaler(
...
+    } catch (Exception e) {
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java]: can this just be `java.lang.IllegalArgumentException?

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java [hunk 3]
```diff
@@ -70,6 +72,13 @@ public EC2AutoScaler(
...
+      log.warn(String.format("Attempting to set an invalid region: %s", envConfig.getRegion()), e);
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java]: mmx logger should take
`log.warn(e, "Attempting to set an invalid region: %s", envConfig.getRegion());`

File: indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerTest.java [hunk 4]
```diff
@@ -56,6 +56,7 @@
   private static final String AMI_ID = "dummy";
   private static final String INSTANCE_ID = "theInstance";
   public static final EC2EnvironmentConfig ENV_CONFIG = new EC2EnvironmentConfig(
+      "us-east-1",
```
Comment (drcrallen) [inline indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerTest.java]: Can you add a test for this being null?

File: indexing-service/src/test/java/io/druid/indexing/overlord/setup/WorkerBehaviorConfigTest.java [hunk 5]
```diff
@@ -49,6 +49,7 @@ public void testSerde() throws Exception
             7,
             11,
             new EC2EnvironmentConfig(
+                "us-east-1",
```
Comment (drcrallen) [inline indexing-service/src/test/java/io/druid/indexing/overlord/setup/WorkerBehaviorConfigTest.java]: Same here, need to retain a test for legacy configs where this is null.

**Comment by drcrallen** [comment]
Hi @nicktylah  thanks for the contrib! would you mind filling out the cla at http://druid.io/community/cla.html ?

**Comment by nicktylah** [comment]
Absolutely! Thanks for the quick review and comments! We love druid at my
company already and are working on getting it in production ASAP.

File: indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerTest.java [hunk 6]
```diff
@@ -72,6 +83,7 @@
   public void setUp() throws Exception
   {
     amazonEC2Client = EasyMock.createMock(AmazonEC2Client.class);
+    mockAmazonEC2Client = new EC2MockClient();
```
Comment (nishantmonu51) [inline indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerTest.java]: can you use EasyMock instead of creating a new class ? 
It doesn't seem to do anything custom else except get/set region.

**Comment by fjy** [comment]
:+1:

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java [hunk 7]
```diff
@@ -70,6 +71,15 @@ public EC2AutoScaler(
     this.envConfig = envConfig;
     this.amazonEC2Client = amazonEC2Client;
     this.config = config;
+
+    final String regionName = envConfig.getRegion();
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java]: per docs in AmazonEC2:


```<b>This method is not threadsafe. A region should be configured when the
     * client is created and before any service requests are made. Changing it
     * afterwards creates inevitable race conditions for any service requests in
     * transit or retrying.</b>```


Can this be moved to `io.druid.guice.AWSModule#getEc2Client`?

**Comment by drcrallen** [comment]
@nicktylah I think this is pretty close to ready to merge in. Can you move the setRegion call (if a region is specified) into the AmazonEC2 provider method? After that I think we're good.

File: indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerSerdeTest.java [hunk 8]
```diff
@@ -28,13 +29,56 @@
...
 {
+  private AmazonEC2Client amazonEC2Client = EasyMock.createMock(AmazonEC2Client.class);
+  private InjectableValues mockInjectableValues = new InjectableValues()
+  {
+    @Override
+    public Object findInjectableValue(
+        Object valueId, DeserializationContext ctxt, BeanProperty forProperty, Object beanInstance
+    )
+    {
+      if (valueId == "com.amazonaws.services.ec2.AmazonEC2") {
```
Comment (drcrallen) [inline indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerSerdeTest.java]: Can you make this `"com.amazonaws.services.ec2.AmazonEC2".equals(valueId)`?

File: indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerSerdeTest.java [hunk 9]
```diff
@@ -28,13 +29,56 @@
...
+        return amazonEC2Client;
+      } else {
+        return null;
```
Comment (drcrallen) [inline indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerSerdeTest.java]: Is there anything else that gets injected? can this just throw an error if something else is requested?

File: aws-common/src/main/java/io/druid/common/aws/ec2/EC2EnvironmentConfig.java [hunk 10]
```diff
@@ -17,7 +17,7 @@
  * under the License.
  */
 
-package io.druid.indexing.overlord.autoscaling.ec2;
+package io.druid.common.aws.ec2;
```
Comment (drcrallen) [inline aws-common/src/main/java/io/druid/common/aws/ec2/EC2EnvironmentConfig.java]: Did this need to move for some reason?

File: server/src/main/java/io/druid/guice/EC2RegionSetter.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+package io.druid.guice;
+
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.ec2.AmazonEC2Client;
+
+/**
+ * Responsible for setting the region on our injected AmazonEC2Client in a thread-safe manner.
+ * Abstracted for testing purposes.
+ */
+public class EC2RegionSetter {
```
Comment (drcrallen) [inline server/src/main/java/io/druid/guice/EC2RegionSetter.java]: Why is this needed?

File: server/src/main/java/io/druid/guice/AWSModule.java [hunk 12]
```diff
@@ -47,8 +51,21 @@ public AWSCredentialsProvider getAWSCredentialsProvider(final AWSCredentialsConf
 
   @Provides
   @LazySingleton
-  public AmazonEC2 getEc2Client(AWSCredentialsProvider credentials)
+  public AmazonEC2Client getEC2Client(
+      AWSCredentialsProvider credentials,
+      EC2RegionSetter ec2RegionSetter,
+      @JsonProperty("envConfig") EC2EnvironmentConfig envConfig)
```
Comment (nicktylah) [inline server/src/main/java/io/druid/guice/AWSModule.java]: I need some guidance getting this environment config into the `AWSModule`. I'm sure it shows -- but I'm a absolute `guice` novice...

File: server/src/main/java/io/druid/guice/EC2RegionSetter.java (hunk 11 above)
Comment (nicktylah) [inline server/src/main/java/io/druid/guice/EC2RegionSetter.java]: Just for testing purposes -- so that the `AWSModule` test do its assertions properly

File: aws-common/src/main/java/io/druid/common/aws/ec2/EC2EnvironmentConfig.java (hunk 10 above)
Comment (nicktylah) [inline aws-common/src/main/java/io/druid/common/aws/ec2/EC2EnvironmentConfig.java]: `AWSModule` needs to access the `getRegion` method on the `EC2EnvironmentConfig` class to set region appropriately -- I figured it was better to have a dependency from `indexing-service` to `aws-common` than from `aws-common` to `indexing-service`

**Comment by nicktylah** [comment]
In truth, I think this PR made more sense when `setRegion` was called in the constructor of `EC2Autoscaler` -- and I can't imagine a problem with thread safety since that line should (as far as I can tell) always run before any usage of the `amazonEC2Client` in the `EC2Autoscaler`

**Comment by drcrallen** [comment]
I looked into it a bit and the setting needed is really far removed from where it needs to be set (in the constructor for the client). And having a region is not an autoscaler specific thing, which is I think where the muddling comes in.

Would it solve your use case if the EC2 client was specific for the global EC2 client?

**Comment by nicktylah** [comment]
In theory, I think it makes total sense for the region to be set in the provider method (`AWSModule`). It's unfortunate that the region config itself is tucked so far away in the indexing service package. If I could find the binding for the `EC2EnvironmentConfig` and play around with that I'm sure I could set it correctly, I'm just struggling to get that piece of information

**Comment by xvrl** [comment]
@nicktylah is there any reason why region needs to be part of the autoscaler config? I can't really think of a reason why someone would want to run the overlord in a different region than the middlemanagers.
In my opinion it would be much simpler to set the region in the runtime properties along with other aws settings and have the autoscaler use the same region as everything else. Please correct me if I am not unerstanding your use-case well.

**Comment by nicktylah** [comment]
@xvl The problem I'm running into is that the aws sdk sets a default region of `us-west-1` if nothing is specified. All my instances run in `us-west-2`, so I get this error in my overlord logs:
```
2016-03-03T04:26:08,645 ERROR [ScalingExec--0] io.druid.indexing.overlord.autoscaling.ec2.EC2AutoScaler - Unable to provision any EC2 instances.
com.amazonaws.AmazonServiceException: Invalid availability zone: [us-west-2b] (Service: AmazonEC2; Status Code: 400; Error Code: InvalidParameterValue; Request ID: f0fdb737-f035-463f-8c05-e78e40a048c6)
```
According to the aws docs, 
> Note
The AWS SDK for Java uses us-east-1 as the default region if you do not specify a region in your code. However, the AWS Management Console uses us-west-2 as its default. Therefore, when using the AWS Management Console in conjunction with your development, be sure to specify the same region in both your code and the console.

**Comment by nicktylah** [comment]
So it looks like the integration tests are failing because of the dependency I've introduced from `server` -> `indexing-service`. This was necessary to introduce the `ResourceManagementSchedulerConfig` class to the `AWSModule`. Do you guys have a recommendation for how to resolve this?

File: server/src/main/java/io/druid/guice/AWSModule.java [hunk 13]
```diff
@@ -47,8 +50,17 @@ public AWSCredentialsProvider getAWSCredentialsProvider(final AWSCredentialsConf
 
   @Provides
   @LazySingleton
-  public AmazonEC2 getEc2Client(AWSCredentialsProvider credentials)
+  public AmazonEC2 getEc2Client(
+      AWSCredentialsProvider credentials,
+      ResourceManagementSchedulerConfig resourceManagementSchedulerConfig)
```
Comment (xvrl) [inline server/src/main/java/io/druid/guice/AWSModule.java]: we cannot make the AWS module dependent on indexing service classes, it should be generic enough that other druid nodes can use it as well. Not every druid node will have a proper ResourceManagementSchedulerConfig available for injection.

File: server/src/main/java/io/druid/indexing/overlord/autoscaling/ResourceManagementSchedulerConfig.java [hunk 14]
```diff
@@ -39,6 +42,11 @@
   @JsonProperty
   private DateTime originTime = new DateTime("2012-01-01T00:55:00.000Z");
 
+  public String getRegion()
```
Comment (xvrl) [inline server/src/main/java/io/druid/indexing/overlord/autoscaling/ResourceManagementSchedulerConfig.java]: We should not introduce AWS-specific concerns in `ResourceManagementSchedulerConfig`. Autoscaling is implementation agnostic, and EC2 is only one possible implementation of auto-scaling. Anything specific to EC2 should stay in the EC2-specific auto-scaling classes.

File: server/src/main/java/io/druid/indexing/overlord/autoscaling/ResourceManagementSchedulerConfig.java [hunk 15]
```diff
@@ -28,6 +28,9 @@
 public class ResourceManagementSchedulerConfig
```
Comment (xvrl) [inline server/src/main/java/io/druid/indexing/overlord/autoscaling/ResourceManagementSchedulerConfig.java]: we shouldn't have to move this class from indexing-service to server. There is no reason why all druid services would need to depend on this class. It is very much an indexing-service specific configuration.

**Comment by xvrl** [comment]
@nicktylah I think we're going to have to rethink a bit more how we want to handle different aws regions. We may also want to consider other things, such as S3 regions to make this a more comprehensive change.
